### ISSUE TITLE
about face tracking 

### ISSUE THREAD

**Issue by mostafa610** [YuanxunLu/LiveSpeechPortraits#49]
first of all thank you so much for your marvelous work.
second, regarding face tracking why do use it why don't you just extract the landmarks from every frame by the landmark detector 
thanks in advance

**Comment by mostafa610** [comment]
and how did the file mean_pts3d.npy created 
is that the mean of this point across the dataset ?

**Comment by mostafa610** [comment]
> 

it is the average of landmarks of all video frames of the target person

**Comment by YuanxunLu** [comment]
1. Using 3d landmarks obtained by face tracking has several advantages over directly using detected 2D landmarks. It helps disentangle the camera parameters, head pose, and facial movements, which allow explicit control over these parameters while using 2D landmarks can't do it. Besides, it is much easier for networks to learn normalized facial movements (in 3D space) than entangled landmarks, generating more accurate results.
2. 'mean_pts3d.npy' should be the mean 3d landmarks of the target person in the training set. The network learns the displacements instead of the absolute locations.

Hope the above helps.

**Comment by mostafa610** [comment]
thank you so much for your reply

please i have another question 
![q](https://user-images.githubusercontent.com/49750203/159441962-deed920c-03ab-40e1-b4aa-dd1fb607a9ab.png)
 here the audio_features: [b, T, ndim]
from the paper i understand that T = 240 is the batch size (number of consecutive frames ) and ndim is 512 
i don't get what is b

**Comment by mostafa610** [comment]
thank you so much for your reply it really helped

I have another question about sequence length in the code you define sequence length to be 240 
in the paper 240 is the batch size (𝑇 = 240 represents the number of consecutive frames sent to
the model at each iteration)  that means it is the batch size ??
---------------------------------------------------------------------------------
If 240 is sequence length then audio features sent during training will be 
[b,T,ndim] = 32(batch size) and 240(seq_length) and 512 (ndim)
----------------------------------------------------------------------------------
or do you mean that you will send batches of one each batch contain sample with 240 seq_length 

thanks in advance

**Comment by YuanxunLu** [comment]
Of course the latter one. LSTM is a kind of RNN network, and it should take sequential data as input. 240 frames equal to 4 seconds under the 60 FPS setting. 

Batch_size means for each forward pass, how many *batches* of sequential data (240 frames data) are sent.

**Comment by mostafa610** [comment]
thank you so much i don't know how to thank you you really helped me !

**Comment by YuanxunLu** [comment]
First, LSTM takes sequential data as input and its output is also sequential, therefore T frames input results in T frames output. Please carefully check the definition of LSTM networks. During training, we use 4 seconds as the length while during the test there is no length limitation.

Secondly, the audio2mouth network learns the displacements.

Thirdly, *frame*2* is because the APC feature frame is half of the 1/60. Please check the paper for details.

**Comment by mostafa610** [comment]
thank you for your replays,
I am wondering do you know any open source algorithm for face tracking as i can't find one to produce the same output of your paper.
thanks in advance

**Comment by YuanxunLu** [comment]
Any parametric monocular face reconstruction method would be an alternative, like FaceScape, DECA, 3DDFA_v2, etc.

**Comment by yuxx0218** [comment]
> Any parametric monocular face reconstruction method would be an alternative, like FaceScape, DECA, 3DDFA_v2, etc.

What the method did you use? could you please upload the code?
